src/cli/topical_doc.rs

```diff
@@ -0,0 +1,139 @@
+use crate::errors::*;
+use std::ffi::OsString;
+use std::fs;
+use std::path::{Path, PathBuf};
+
+struct DocData<'a> {
+    topic: &'a str,
+    subtopic: &'a str,
+    root: &'a Path,
+}
+
+fn no_document(topic: &str) -> Result<PathBuf> {
+    Err(format!("No document for '{}'", topic).into())
+}
+
+fn index_html(doc: &DocData, wpath: &Path) -> Option<PathBuf> {
+    let indexhtml = wpath.join("index.html");
+    match &doc.root.join(&indexhtml).exists() {
+        true => Some(indexhtml),
+        false => None,
+    }
+}
+
+fn dir_into_vec(dir: &PathBuf) -> Result<Vec<OsString>> {
+    let entries = fs::read_dir(dir).chain_err(|| format!("Opening directory {:?}", dir))?;
+    let mut v = Vec::new();
+    for entry in entries {
+        let entry = entry?;
+        v.push(entry.file_name());
+    }
+    Ok(v)
+}
+
+fn search_path(doc: &DocData, wpath: &Path, keywords: &[&str]) -> Result<PathBuf> {
+    let dir = &doc.root.join(&wpath);
+    if dir.is_dir() {
+        let entries = dir_into_vec(dir)?;
+        for k in keywords {
+            let filename = &format!("{}.{}.html", k, doc.subtopic);
+            if entries.contains(&OsString::from(filename)) {
+                return Ok(dir.join(filename));
+            }
+        }
+        no_document(doc.topic)
+    } else {
+        no_document(doc.topic)
+    }
+}
+
+pub fn local_path(root: &Path, topic: &str) -> Result<String> {
+    let keywords_top = ["macro", "keyword", "primitive"];
+    let keywords_mod = ["fn", "struct", "trait", "enum", "type", "constant"];
+
+    let topic_vec: Vec<&str> = topic.split("::").collect();
+    let work_path = topic_vec.iter().fold(PathBuf::new(), |acc, e| acc.join(e));
+
+    let doc = DocData {
+        topic: &topic,
+        subtopic: topic_vec[topic_vec.len() - 1],
+        root,
+    };
+
+    /**************************
+
```

Comment:
It may be worth referencing your `tests/mock/topical_doc_data.rs` file here to say that if additional functionality is added, it should be added to the tests so that it can be verified.  That way we'll keep support for new doc sections properly supported if `std` is split up further.